### DISCUSSION TITLE
RandCropByPosNegLabeld  self.center mix for multi-process and dataset different shape

### DISCUSSION THREAD

**Discussion by fgbxx** [Project-MONAI/MONAI#3215] [Q&A]
hi, all 
  I found a bug when I use multi-process for different shape dataset.
  the transform is global , and the self.center will be changed during different processes,  so change it to a local parameter in call function is better solution I think.

Best, 
Gang in Beijing

**Comment by wyli** [comment]
If the inputs have different shapes, perhaps you should create separate `RandCropByPosNegLabeld` instances and set different `keys` respectively. I'm moving this to the discussion tab.

**Reply by rijobro** [reply]
I think the point is that in one thread you might generate some centres, and in the second thread you generate some others. When you then go to apply the generated centres in the first thread, you might use the centres generated by the second thread. To me this seems like a valid issue -- we shouldn't be setting any `self` variables in the `randomize` methods of our transforms (unless they're marked `ThreadUnsafe`).

**Reply by fgbxx** [reply]
> I think the point is that in one thread you might generate some centres, and in the second thread you generate some others. When you then go to apply the generated centres in the first thread, you might use the centres generated by the second thread. To me this seems like a valid issue -- we shouldn't be setting any `self` variables in the `randomize` methods of our transforms (unless they're marked `ThreadUnsafe`).

yes, you got my idea. I think should consider threadsafe in tansform. 

For medical image, different shape is very common ,  we use RandCropByPosNegLabeld to get same patch size for trainning, it is very resonable, so support different original volumn is  neccessary.

Look forward to fix.

Best, 
Gang @BeiJing

**Reply by rijobro** [reply]
Hi @fgbxx, I drafted a PR for this, but following some [discussion](https://github.com/Project-MONAI/MONAI/pull/3217#discussion_r739375269), I realised that our random transforms are inherently thread unsafe (and therefore inherit from `ThreadUnsafe`).

@wyli @Nic-Ma could you weigh in here -- if all of our random transforms are `ThreadUnsafe`, does that not make them dangerous to use with `CacheDataset`, `PersistentDataset` or even `DataLoader(num_workers>1)`? This seems like a large use case. Perhaps I've misunderstood something.

**Reply by Nic-Ma** [reply]
Hi @rijobro ,

As you know, PyTorch num_workers are multi-processing instead of multi-threads, I think it's OK to run Thread unsafe code.

Thanks.

**Reply by tmathai** [reply]
Hi @Nic-Ma and @rijobro. I am facing the same bug in my segmentation project while using MONAI (version: 0.8.0+97.gc2b64593) with `RandCropByPosNegLabeld`. The pipeline I follow is:

1. Load my nifti volume and labels
2. Add spatial padding (via `SpatialPadd`) of (256, 256, 32) to make them divisible by the UNet layers
3. Use `RandCropByPosNegLabeld` with the same `spatial_size`. 

However, when I plot the cropped labels over cropped slices of dims (256, 256, 32), the labels are misaligned and sometimes completely off. I checked the dimensions of the nifti volumes and masks in my project and they match. I also checked without `RandCropByPosNegLabeld` Any suggestions to fix this issue? Are there any updates to this request?

**Reply by rijobro** [reply]
Can you show us the code? Something like:

```python
keys = ["image", "label"]
transform = Compose([
    LoadImaged(keys),
    SpatialPadd(keys, ...),
    RandCropByPosNegLabeld(keys, ....),
])
data = {"image": filename, "label": filename2}
data = transform(data)
# visualise
...
```

**Reply by tmathai** [reply]
@Nic-Ma @rijobro Sure, sorry it took a while. Had to make sure that this was indeed a problem. So, I did a small unit test on the Spleen dataset with both v0.8 and v0.7, and noticed that the cropped labels were off when I plotted them on cropped images. Also I wrote my own random cropper, but that didn't work because I have to follow the prototype of transforms in the `__call__` method of `RandCropByPosNegLabeld` and additional fields added after transforming the data implemented in the code (e.g. `self.push_transform`). Is there a bug in my code that I am missing, or is there something inherently wrong with the cropper?  Any suggestions are welcome to help fix this issue? Here is my MWE:

## Gather Spleen data
```
# directory to save spleen (medical decathalon) dataset into
root_dir = '/home/tsm/Data'
print('data_dir:', root_dir)

model_dir = '/home/tsm/Models/LN_seg/'
print('model_dir:', model_dir)

data_dir = os.path.join(root_dir, "Task09_Spleen")

train_images = sorted(
    glob.glob(os.path.join(data_dir, "imagesTr", "*.nii.gz")))
train_labels = sorted(
    glob.glob(os.path.join(data_dir, "labelsTr", "*.nii.gz")))
data_dicts = [
    {"image": image_name, "label": label_name}
    for image_name, label_name in zip(train_images, train_labels)
]
train_files, val_files = data_dicts[:-9], data_dicts[-9:]

print()
# print(train_files)
print(len(train_files))

print()
# print(val_files)
print(len(val_files))

set_determinism(seed=0)
```

## Some validation transforms 

```
val_transforms = Compose(
    [
        LoadImaged(keys=["image", "label"]),
        EnsureChannelFirstd(keys=["image", "label"]),
        Orientationd(keys=["image", "label"], axcodes="RAS"),
        ScaleIntensityRanged(
            keys=["image"], a_min=-57, a_max=164,
            b_min=0.0, b_max=1.0, clip=True,
        ),
        RandCropByPosNegLabeld(
            keys=["image", "label"],
            label_key="label",
            spatial_size=(256, 256, 32),
            pos=1,
            neg=1,
            num_samples=4,
            image_key="image",
            image_threshold=0,
        ),
        EnsureTyped(keys=["image", "label"]),
    ]
)
```

## Cache data

```
val_ds = CacheDataset(
                            data = val_files, 
                            transform = val_transforms,
                            cache_rate = 1, 
                            num_workers = 4
                        )
# val_ds = Dataset(data=val_files, transform=val_transforms)
val_loader = DataLoader(val_ds, batch_size=1, shuffle=False, num_workers=4)
 
```

## Plot data after post-validation transforms to check training + val Dice metric  

```
device = torch.device("cuda:3")

## v0.7
# post_label = Compose([EnsureType(), AsDiscrete(to_onehot=True, num_classes = 2)])
## v0.8
post_label = Compose([EnsureType(), AsDiscrete(to_onehot=2)])
post_pred = Compose([Activations(sigmoid=True), AsDiscrete(argmax=True)])

with torch.no_grad():
    
    print('len val_loader (num valid scans):', len(val_loader))
    
    # select one image to evaluate and visualize the model output
    scan_idx = 6
    #sample_idx = 0
    
    ## with myOwnCropper
#     num_samples = len(val_ds[scan_idx]['image'])
#     ## with RandCropByPosNegLabeld
    num_samples = len(val_ds[scan_idx])
    print('num_samples:', num_samples)
    
    for sample_idx in range(1):
    
        print()

        val_inputs = val_ds[scan_idx][sample_idx]["image"].unsqueeze(0).to(device)
        val_labels = val_ds[scan_idx][sample_idx]["label"].unsqueeze(0).to(device)
#         ## without RandCropByPosNegLabeld
#         val_inputs = val_ds[scan_idx]["image"].unsqueeze(0).to(device)
#         val_labels = val_ds[scan_idx]["label"].unsqueeze(0).to(device)
        

        print('val_inputs shape:', val_inputs.shape)
        print('val_labels shape:', val_labels.shape)

        # apply transform
        val_labels = [post_label(i) for i in decollate_batch(val_labels)]

        print('post-transform, val_labels shape:', val_labels[0].shape)

        # get the numpy input image
        valid_input = val_inputs[0].detach().cpu().numpy()
        # get the numpy input label
        valid_label = val_labels[0].detach().cpu().numpy()

        print('numpy valid_input shape:', valid_input.shape)
        print('numpy valid_label shape:', valid_label.shape)
        
        # get annotation slices from the cropped volume
        gt_slices = np.nonzero(np.sum(valid_label[1, :, :, :], axis = (0,1)) > 0)[0]

        # number of nifti slices
        num_nifti_slices = valid_input.shape[-1]

        for slice_idx in gt_slices:

            print('slice_idx:', slice_idx)        

            # get image
            imageSlice_to_show = valid_input[0, :, :, slice_idx]

            # get mask 
            maskSlice_to_show = valid_label[1, :, :,slice_idx]

            plt.figure("visualize", (18, 18))
            plt.subplot(1, 4, 1)
            plt.title("image")
            plt.imshow(imageSlice_to_show, cmap="gray")

            plt.subplot(1, 4, 2)
            plt.title("mask")
            plt.imshow(imageSlice_to_show, cmap="gray")
            plt.imshow(maskSlice_to_show, cmap='jet', interpolation='none', alpha=0.5)

            plt.show()
```

### First run output

<img width="537" alt="image" src="https://user-images.githubusercontent.com/11764041/153629045-857a7ac3-bcd7-42c3-8f23-36a0123505d9.png">

### Another run output

<img width="527" alt="image" src="https://user-images.githubusercontent.com/11764041/153629964-56fda564-93ac-4fd1-b527-2cffad5d5b25.png">

**Reply by tmathai** [reply]
@rijobro @Nic-Ma @wyli I'm seeing this odd behavior whilst flipping the volume with `RandFlipd`. This should be basic functionality in a DL pipeline, so I'm at a loss for words to explain this? Any help here would be appreciated. I tried looking into the code, and it seems like randomizing functionality with `set_random_state` and `self.R` might be the culprits.  

```
val_transforms = Compose(
    [
        LoadImaged(keys=["image", "label"]),
        EnsureChannelFirstd(keys=["image", "label"]),
        Orientationd(keys=["image", "label"], axcodes="RAS"),
        ScaleIntensityRanged(
            keys=["image"], a_min=-57, a_max=164,
            b_min=0.0, b_max=1.0, clip=True,
        ),
        RandFlipd(
            keys=["image", "label"],
            prob = 0.5, 
            spatial_axis = 0, 
        ),
        EnsureTyped(keys=["image", "label"]),
    ]
)
```

<img width="539" alt="image" src="https://user-images.githubusercontent.com/11764041/153663182-ecb01abf-9c81-45d5-899d-e2ec2c337ccd.png">

**Reply by wyli** [reply]
perhaps the same as https://github.com/Project-MONAI/MONAI/discussions/2827

**Reply by tmathai** [reply]
> perhaps the same as #2827

@wyli Ah that makes sense, thanks! I just did `curr_data_point = val_ds[scan_idx][sample_idx]` and the results were consistent. However, setting determinism should work according to the thread you posted, but it doesn't. Anyhow, thanks!

**Reply by wyli** [reply]
thanks for confirming!

**Reply by tmathai** [reply]
@wyli Maybe the spleen/brats 3D segmentation tutorial(s) should be updated to reflect this issue of a randomized transform every time a data point from the data loader is called upon? This would prevent others from spending 3 days thinking that there is a bug somewhere 🤣

**Reply by rijobro** [reply]
Hi @tmathai, which line(s) of the tutorial do you find confusing? I don't see anywhere in which the transforms are called twice (once for the image and then once for the label).

**Reply by tmathai** [reply]
@rijobro In the BRATS 3D segmentation tutorial, specifically at box 7 (`In 7`) called `Check data shape and visualize`, there are two instances where data is extracted: 1. `print` statement and in the `plt.imshow`statement. That was where I grabbed the code to extract the data point ( image/label) after transformations and plot the masks over the image, and my code above shows that.

**Reply by rijobro** [reply]
@tmathai thanks for point it out. In this case it works since there are no random validation transforms, but you're right that this should be simplified.

**Reply by rijobro** [reply]
PR here: https://github.com/Project-MONAI/tutorials/pull/549
